### ISSUE TITLE
[fsck-test] 001-bad-file-extent-bytenr: memory leak

### ISSUE THREAD

**Issue by matthiaskrgr** [kdave/btrfs-progs#30]
compiled with ````clang 4.0````
CFLAGS: ````-g3 -O0 -Wall -fno-omit-frame-pointer -fsanitize-address-use-after-scope  -fno-optimize-sibling-calls -fsanitize=address,undefined````
based on master branch / btrfs progs 4.10.1  / 15d37febbce1255c1042764690a5e5375d014616
run command (as root): ````UBSAN_OPTIONS=print_stacktrace=1 make test -k````

[fsck-tests-results.txt](https://github.com/kdave/btrfs-progs/files/853003/fsck-tests-results.txt)

Test showed a memory leak
````
=== Entering /home/matthias/vcs/github/btrfs-progs/tests/fsck-tests/001-bad-file-extent-bytenr
restoring image default_case.img
testing image default_case.img.restored
checking extents
ref mismatch on [12845056 1048576] extent item 1, found 0
Incorrect local backref count on 12845056 root 5 owner 257 offset 0 found 0 wanted 1 back 0x607000000480
Backref disk bytenr does not match extent record, bytenr=12845056, ref bytenr=13744632839234567870
backpointer mismatch on [12845056 1048576]
owner ref check failed [12845056 1048576]
ref mismatch on [755944791 1048576] extent item 0, found 1
Backref 755944791 root 5 owner 257 offset 0 num_refs 0 not found in extent tree
Incorrect local backref count on 755944791 root 5 owner 257 offset 0 found 1 wanted 0 back 0x6070000004f0
backpointer mismatch on [755944791 1048576]
bad extent [755944791, 756993367), type mismatch with chunk
checking free space cache
checking fs roots
root 5 inode 257 errors 1000, some csum missing
ERROR: errors found in fs roots
Checking filesystem on ./default_case.img.restored
UUID: 17442351-31aa-45fa-9503-90fd48874c3e
cache and super generation don't match, space cache will be invalidated
found 2392064 bytes used, error(s) found
total csum bytes: 1024
total tree bytes: 32768
total fs tree bytes: 8192
total extent tree bytes: 4096
btree space waste bytes: 24211
file data blocks allocated: 1310720
 referenced 1310720
############### /home/matthias/vcs/github/btrfs-progs/btrfs check --repair ./default_case.img.restored
checking extents
ref mismatch on [12845056 1048576] extent item 1, found 0
Incorrect local backref count on 12845056 root 5 owner 257 offset 0 found 0 wanted 1 back 0x607000000480
Backref disk bytenr does not match extent record, bytenr=12845056, ref bytenr=13744632839234567870
backpointer mismatch on [12845056 1048576]
owner ref check failed [12845056 1048576]
attempting to repair backref discrepency for bytenr 12845056
Fixed 0 roots.
checking free space cache
checking fs roots
checking csums
checking root refs
enabling repair mode
Checking filesystem on ./default_case.img.restored
UUID: 17442351-31aa-45fa-9503-90fd48874c3e
cache and super generation don't match, space cache will be invalidated
found 3735552 bytes used, no error found
total csum bytes: 2048
total tree bytes: 65536
total fs tree bytes: 16384
total extent tree bytes: 8192
btree space waste bytes: 48422
file data blocks allocated: 2621440
 referenced 2621440
############### /home/matthias/vcs/github/btrfs-progs/btrfs check ./default_case.img.restored
checking extents
checking free space cache
checking fs roots
checking csums
checking root refs
Checking filesystem on ./default_case.img.restored
UUID: 17442351-31aa-45fa-9503-90fd48874c3e
cache and super generation don't match, space cache will be invalidated
found 1343488 bytes used, no error found
total csum bytes: 1024
total tree bytes: 32768
total fs tree bytes: 8192
total extent tree bytes: 4096
btree space waste bytes: 24211
file data blocks allocated: 1310720
 referenced 1310720
=== Entering /home/matthias/vcs/github/btrfs-progs/tests/fsck-tests/002-bad-transid
restoring image default_case.img
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
Ignoring transid failure
testing image default_case.img.restored
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
Ignoring transid failure
checking extents
checking free space cache
checking fs roots
checking csums
checking root refs
ERROR: transid errors in file system
Checking filesystem on ./default_case.img.restored
UUID: e0da4603-3bc7-4678-8e1d-d4913cff33e3
cache and super generation don't match, space cache will be invalidated
found 1343488 bytes used, error(s) found
total csum bytes: 1024
total tree bytes: 32768
total fs tree bytes: 8192
total extent tree bytes: 4096
btree space waste bytes: 24211
file data blocks allocated: 1310720
 referenced 1310720
############### /home/matthias/vcs/github/btrfs-progs/btrfs check --repair ./default_case.img.restored
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
parent transid verify failed on 29360128 wanted 9 found 755944791
Ignoring transid failure
checking extents
Fixed 0 roots.
checking free space cache
checking fs roots
checking csums
checking root refs
extent buffer leak: start 29360128 len 4096
enabling repair mode
Checking filesystem on ./default_case.img.restored
UUID: e0da4603-3bc7-4678-8e1d-d4913cff33e3
cache and super generation don't match, space cache will be invalidated
Recowing metadata block 29360128
found 1343488 bytes used, no error found
total csum bytes: 1024
total tree bytes: 32768
total fs tree bytes: 8192
total extent tree bytes: 4096
btree space waste bytes: 24211
file data blocks allocated: 1310720
 referenced 1310720
############### /home/matthias/vcs/github/btrfs-progs/btrfs check ./default_case.img.restored
checking extents
checking free space cache
checking fs roots
checking csums
checking root refs
Checking filesystem on ./default_case.img.restored
UUID: e0da4603-3bc7-4678-8e1d-d4913cff33e3
cache and super generation don't match, space cache will be invalidated
found 1343488 bytes used, no error found
total csum bytes: 1024
total tree bytes: 32768
total fs tree bytes: 8192
total extent tree bytes: 4096
btree space waste bytes: 24211
file data blocks allocated: 1310720
 referenced 1310720
=== Entering /home/matthias/vcs/github/btrfs-progs/tests/fsck-tests/003-shift-offsets
restoring image default_case.img
testing image default_case.img.restored
checking extents
incorrect offsets 15018 9387
bad block 29360128
ERROR: errors found in extent allocation tree or chunk allocation
checking free space cache
checking fs roots
incorrect offsets 15018 9387
ERROR: errors found in fs roots

=================================================================
==26624==ERROR: LeakSanitizer: detected memory leaks

Indirect leak of 768 byte(s) in 4 object(s) allocated from:
    #0 0x4ca51c in malloc /home/matthias/LLVM/LLVM_4_0/stage_2/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:66:3
    #1 0x7df0b1 in add_extent_rec_nolookup /home/matthias/vcs/github/btrfs-progs/cmds-check.c:6032:8
    #2 0x7dd409 in add_extent_rec /home/matthias/vcs/github/btrfs-progs/cmds-check.c:6174:8
    #3 0x7e7c7b in process_extent_item /home/matthias/vcs/github/btrfs-progs/cmds-check.c:6818:2
    #4 0x7d658f in run_next_block /home/matthias/vcs/github/btrfs-progs/cmds-check.c:7710:5
    #5 0x7c995b in deal_root_from_list /home/matthias/vcs/github/btrfs-progs/cmds-check.c:9751:10
    #6 0x7954e8 in check_chunks_and_extents /home/matthias/vcs/github/btrfs-progs/cmds-check.c:9918:8
    #7 0x78b12d in cmd_check /home/matthias/vcs/github/btrfs-progs/cmds-check.c:12961:9
    #8 0x4f8dd5 in main /home/matthias/vcs/github/btrfs-progs/btrfs.c:246:8
    #9 0x7f697628a400 in __libc_start_main /usr/src/debug/glibc-2.24-33-ge9e69e4/csu/../csu/libc-start.c:289

Indirect leak of 216 byte(s) in 3 object(s) allocated from:
    #0 0x4ca51c in malloc /home/matthias/LLVM/LLVM_4_0/stage_2/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:66:3
    #1 0x7c8349 in add_root_item_to_list /home/matthias/vcs/github/btrfs-progs/cmds-check.c:9683:11
    #2 0x794efc in check_chunks_and_extents /home/matthias/vcs/github/btrfs-progs/cmds-check.c:9886:11
    #3 0x78b12d in cmd_check /home/matthias/vcs/github/btrfs-progs/cmds-check.c:12961:9
    #4 0x4f8dd5 in main /home/matthias/vcs/github/btrfs-progs/btrfs.c:246:8
    #5 0x7f697628a400 in __libc_start_main /usr/src/debug/glibc-2.24-33-ge9e69e4/csu/../csu/libc-start.c:289

Indirect leak of 128 byte(s) in 4 object(s) allocated from:
    #0 0x4ca51c in malloc /home/matthias/LLVM/LLVM_4_0/stage_2/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:66:3
    #1 0x7e1f05 in alloc_tree_backref /home/matthias/vcs/github/btrfs-progs/cmds-check.c:5886:29
    #2 0x7ddb48 in add_tree_backref /home/matthias/vcs/github/btrfs-progs/cmds-check.c:6216:10
    #3 0x7e7f8b in process_extent_item /home/matthias/vcs/github/btrfs-progs/cmds-check.c:6832:10
    #4 0x7d658f in run_next_block /home/matthias/vcs/github/btrfs-progs/cmds-check.c:7710:5
    #5 0x7c995b in deal_root_from_list /home/matthias/vcs/github/btrfs-progs/cmds-check.c:9751:10
    #6 0x7954e8 in check_chunks_and_extents /home/matthias/vcs/github/btrfs-progs/cmds-check.c:9918:8
    #7 0x78b12d in cmd_check /home/matthias/vcs/github/btrfs-progs/cmds-check.c:12961:9
    #8 0x4f8dd5 in main /home/matthias/vcs/github/btrfs-progs/btrfs.c:246:8
    #9 0x7f697628a400 in __libc_start_main /usr/src/debug/glibc-2.24-33-ge9e69e4/csu/../csu/libc-start.c:289

SUMMARY: AddressSanitizer: 1112 byte(s) leaked in 11 allocation(s).
############### /home/matthias/vcs/github/btrfs-progs/btrfs check --repair ./default_case.img.restored
checking extents
incorrect offsets 15018 9387
Fixed 0 roots.
checking free space cache
checking fs roots
checking csums
checking root refs
enabling repair mode
Checking filesystem on ./default_case.img.restored
UUID: 96ab8688-d6d5-44b0-849e-63720c2c79db
Shifting item nr 28 by 5631 bytes in block 29376512
Shifting item nr 29 by 5631 bytes in block 29376512
Shifting item nr 30 by 5631 bytes in block 29376512
Shifting item nr 31 by 5631 bytes in block 29376512
Shifting item nr 32 by 5631 bytes in block 29376512
Shifting item nr 33 by 5631 bytes in block 29376512
Shifting item nr 34 by 5631 bytes in block 29376512
Shifting item nr 35 by 5631 bytes in block 29376512
Shifting item nr 36 by 5631 bytes in block 29376512
Shifting item nr 37 by 5631 bytes in block 29376512
Shifting item nr 38 by 5631 bytes in block 29376512
Shifting item nr 39 by 5631 bytes in block 29376512
Shifting item nr 40 by 5631 bytes in block 29376512
Shifting item nr 41 by 5631 bytes in block 29376512
Shifting item nr 42 by 5631 bytes in block 29376512
Shifting item nr 43 by 5631 bytes in block 29376512
Shifting item nr 44 by 5631 bytes in block 29376512
Shifting item nr 45 by 5631 bytes in block 29376512
Shifting item nr 46 by 5631 bytes in block 29376512
Shifting item nr 47 by 5631 bytes in block 29376512
Shifting item nr 48 by 5631 bytes in block 29376512
Shifting item nr 49 by 5631 bytes in block 29376512
Shifting item nr 50 by 5631 bytes in block 29376512
Shifting item nr 51 by 5631 bytes in block 29376512
Shifting item nr 52 by 5631 bytes in block 29376512
Shifting item nr 53 by 5631 bytes in block 29376512
Shifting item nr 54 by 5631 bytes in block 29376512
Shifting item nr 55 by 5631 bytes in block 29376512
Shifting item nr 56 by 5631 bytes in block 29376512
cache and super generation don't match, space cache will be invalidated
found 786432 bytes used, no error found
total csum bytes: 0
total tree bytes: 196608
total fs tree bytes: 32768
total extent tree bytes: 32768
btree space waste bytes: 178837
file data blocks allocated: 524288
 referenced 524288
############### /home/matthias/vcs/github/btrfs-progs/btrfs check ./default_case.img.restored
checking extents
checking free space cache
checking fs roots
checking csums
checking root refs
Checking filesystem on ./default_case.img.restored
UUID: 96ab8688-d6d5-44b0-849e-63720c2c79db
cache and super generation don't match, space cache will be invalidated
found 393216 bytes used, no error found
total csum bytes: 0
total tree bytes: 131072
total fs tree bytes: 32768
total extent tree bytes: 16384
btree space waste bytes: 119363
file data blocks allocated: 262144
 referenced 262144
`````

**Comment by adam900710** [comment]
I know this is necrobumping, but this is already fixed by v4.15.1, with various asan fixes.

**Comment by kdave** [comment]
Ok thanks, let's close it.
